app/models/peer_review.rb

```diff
@@ -69,16 +69,61 @@ def self.assign(reviewer_groups, reviewee_groups)
   end
 
   def self.unassign(selected_reviewee_group_ids, reviewers_to_remove_from_reviewees_map)
+    delete_all_reviews = true
+    deleted_count = 0
+    undeleted_reviews = []
+
+    # When an entire row is selected for unassignment
+    [selected_reviewee_group_ids].each do |reviewee_id| # process each reviewee when a whole row is selected
+      peer_reviews = PeerReview.joins(result: :submission)  # query all peer reviews in a selected row
+                               .where(submissions: { grouping_id: reviewee_id })
+
+      peer_reviews.each do |peer_review|
+        reviewer_id = peer_review.reviewer.id
+        reviewers_to_remove_from_reviewees_map[reviewee_id] ||= {} # initialize the hash as empty if it doesn't exist
+        reviewers_to_remove_from_reviewees_map[reviewee_id][reviewer_id.to_s] = 'true' # mark the reviewer to be removed
+      end
+    end
+
     # First do specific unassigning.
     reviewers_to_remove_from_reviewees_map.each do |reviewee_id, reviewer_id_to_bool|
       reviewer_id_to_bool.each_key do |reviewer_id|
         reviewee_group = Grouping.find_by(id: reviewee_id)
         reviewer_group = Grouping.find_by(id: reviewer_id)
-        self.delete_peer_review_between(reviewer_group, reviewee_group)
+
+        unless reviewee_group.nil? || reviewer_group.nil?
+          peer_review = reviewer_group.review_for(reviewee_group)
+          unless peer_review.nil?
+            if peer_review.has_marks_or_annotations?
```

Comment:
This check (`peer_review.has_marks_or_annotations?`) is good, but you should reorganize as follows: instead of adding it here, add it as a [`before_destroy` callback](https://guides.rubyonrails.org/active_record_callbacks.html#destroying-an-object). The `delete_peer_review_between` method calls `destroy`, and will trigger the callback. This is more robust because it will affect everywhere that a peer review is attempted to be destroyed, not just this particular method.

(It's possible to also skip callbacks, in case we ever want to enable that in the future.)